### ISSUE TITLE
d.base_path sometimes empty

### ISSUE THREAD

**Issue by chros73** [rakshasa/rtorrent#448]
Since I couldn't reopen this old issue #80 (d.get_base_path on closed torrents) hence the new one.

> A while back I ran into an issue where d.get_base_path would be an empty string when torrents were closed (I believe in between sessions). Ran into this again, and looking around the source I see it returns an empty string if the download's file_list is empty. So I guess the file_list isn't repopulated when rtorrent restarts and a torrent is closed. This seems like straightforward, intended behavior. In fact, if I start a torrent again I can see the base_path again.

I observed the very same effect with v0.9.6. Steps to reproduce this:
- stop a finished/unfinished torrent
- stop rtorrent then start it again

That can be a real problem during deleting/moving torrents.
So, what's the best way to determine the actual full path for a torrent in every case?

**Comment by pyroscope** [comment]
"directory" is always available, and then you need to add "name" yourself for single-file torrents. This changed somewhere around 0.9.2 or 0.8.9, earlier base_path was always available.

**Comment by chros73** [comment]
Thanks You for the suggestion. Here it is what @blaenk said: https://github.com/rakshasa/rtorrent/blob/master/src/command_download.cc#L72

Now the "only" question remains how to determine perfectly whether a torrent contains only a 'single file' and NOT a 'single file in a directory' :) `d.size_files` reports `1` for both cases so it's not good for this.

**Comment by chros73** [comment]
I just noticed that similarly `d.base_filename` is also empty in this case.
We can replace it with `d.name`.

Thanks for mentioning (in chat) the `d.is_multi_file` method to determine whether a torrent contains only a single file.
So this is how it should look like:

```
# gets the full path of data of a torrent (it's a workaround for the possibly empty 'd.base_path' attribute)
method.insert  = d.get_data_full_path, simple, "branch=d.is_multi_file=,cat=$d.directory=,\"cat=$d.directory=,/,$d.name=\""
```

With this new method a simple deletion should look like this (of course additional checks wouldn't hurt):
```
method.insert  = d.remove_data, simple|private, "execute={rm,-rf,--,$d.get_data_full_path=}"
```

(And moving torrents will be a little more complicated thanks to we supposed to overwrite the `d.directory` attribute at first, but it's manageable.)

Thanks for your help!

**Comment by pyroscope** [comment]
You better make damn sure d.name is not empty of "funny" when using it with `rm -rf`. :trollface:

**Comment by chros73** [comment]
Are you kidding me?! :) That was the whole purpose of this change! :)
Have you seen any case where it contains garbage (e.g. empty)? If so, can we reproduce it? (After a short test it looks fine.)

**Comment by pyroscope** [comment]
No I don't, right now. But if it ever does, say goodbye to all the data in your default download location.

The safe way is to use "rm -rf" with multi-file downloads only (i.e. copy that branch to your delete action).

**Comment by chros73** [comment]
> The safest way is rtcontrol --cull, since that adds the files IN the metafile and deletes only those, and cleans up empty dirs left behind.

Interesting.

I went on a different road with "rm -rf": I'm checking before deletion whether a dir or file exists (using the appropriate one on each cases). https://github.com/chros73/rtorrent-ps_setup/blob/master/ubuntu-14.04/home/chros73/.rtorrent.rc#L98
Here the `d.data_full_path_exist` method is the key (it uses another helper method).

So, what do you think about this solution?
I consider the config as done now. :)

But thanks for mentioning this, because I forgot to include a similar check during moving (could have caused damage as well), now I did this as well. :)

Btw, I also applied this very same logic in the Data-directory column in my fork of your beautiful rtorrent-ps (https://github.com/chros73/rtorrent-ps#fork-notes). (I've also uploaded some screenshots: https://github.com/chros73/rtorrent-ps/tree/master/docs/_static/img)

**Comment by pyroscope** [comment]
The probability that any odd path in your filesystem contains the content of your items as sub-paths are near zero. And limiting changes to a "topdir" or whatever severely limits use-cases where people have multiple partitions in /mnt or similar stuff.

**Comment by chros73** [comment]
> The probability that any odd path in your filesystem contains the content of your items as sub-paths are near zero.

Agreed. :) I just wanted to let you know.

> And limiting changes to a "topdir" or whatever severely limits use-cases where people have multiple partitions in /mnt or similar stuff.

They can use symlinks to have a "topdir". That's what I'm using.
But, this is a business decision to make, what we want to accomplish.

I went for the "jailroot" type setup, especially that I know where the deletable data should be (I'm using a carefully updated `d.custom=data_dir` custom variable from the beginning).

> As far as "works equally good" goes, DO download two multi-file items that have the exact same base dir, but differing file lists. Delete one. What happens?

:D You're right. And I haven't even known that we can download different contents into the same base dir until now! I thought we can't. :) (I always move the data into a different dir (category) when download is finished.)
I'll check whether `cull=` can be used in my setup (I'm using `event.download.erased` event to trigger deletion).

Thanks for the explanation!

**Comment by pyroscope** [comment]
You can use cull, but must use --detach or one of rtorrent's *.bg variants, since otherwise rtorrent dead-locks. That might be a problem depending on what your handlers do. You could use https://github.com/chneukirchen/nq or similar to run deletion asynchronous.

rtmv does only same-device mv+ln right now, that will change though.

**Comment by chros73** [comment]
I tried it: I can't use `cull` with `event.download.erased`. Probably the torrent is already "gone" in the meantime when `rtcontrol` tries to modify it. :)

Anyway, thanks for pointing out to this bug, I took a note of it. (It shouldn't happen really often anyway with auto-moving feature. "The probability is near to zero" :D No, it's probably higher though :) )

I leave it as it is for now, it would change everything from the basics. Maybe in the far distant future I'll rewrite the setup to use pyrocore utils for deleting, moving, queuing, rotating torrents when `sweep` feature will be implemented.

Cheers
